### PR TITLE
ARROW-17692: [R] Add support for building with system AWS SDK C++

### PR DESCRIPTION
**Pull request by thisisnic**
This PR uses "pkg-config --static ... arrow" to collect build flags. "pkg-config --static ... arrow" reports suitable build flags that depend on build options and used libraries for Apache Arrow C++. This works with the system AWS SDK C++.

### CODE REVIEW

File: r/configure [hunk 1]
```diff
@@ -255,8 +255,8 @@ if [ $? -eq 0 ]; then
   if arrow_built_with ARROW_S3; then
     PKG_CFLAGS="$PKG_CFLAGS -DARROW_R_WITH_S3"
     if [ "$BUNDLED_LIBS" != "" ]; then
-      # We're depending on openssl/curl from the system, so they're not in the bundled deps
-      BUNDLED_LIBS="$BUNDLED_LIBS -lssl -lcrypto -lcurl"
+      # We're depending on openssl/curl/AWS SDK from the system, so they're not in the bundled deps
+      BUNDLED_LIBS="$BUNDLED_LIBS -lssl -lcrypto -lcurl -laws-cpp-sdk-identity-management -laws-cpp-sdk-sts -laws-cpp-sdk-cognito-identity -laws-cpp-sdk-s3 -laws-cpp-sdk-core -laws-c-event-stream -laws-checksums -laws-c-common"
```
Comment (wjones127) [inline r/configure:259]: This seems fine, but I honestly am very confused what `BUNDLED_LIBS` is supposed to represent (or, why aren't we just adding them to `PKG_LIBS`?). Maybe it's that these are the dependencies of what is in the bundled libraries, so they need to *after* when linking?

Comment (kou) [reply r/configure:259]: 1. We need to add AWS SDK C++ related `-lXXX`s only when Apache Arrow C++ is built with system AWS SDK C++ (If we can require `arrow.pc`, we can implement it simply...)
2. We need to add AWS SDK C++ related `-lXXX`s before `-lssl -lcrypto -lcurl` because AWS SDK C++ depend on `-lssl -lcrypto -lcurl`

Comment (nealrichardson) [reply r/configure:259]: I think we can use pkg-config to detect what libs are required from the system. In fact, we already do on L187. So is the right fix in the arrow.pc file? (Or probably in the pkg-config file that aws-sdk-cpp produces?)

**Comment by kou** [comment]
@github-actions crossbow submit -g r

**Comment by kou** [comment]
@github-actions crossbow submit -g r

**Comment by kou** [comment]
@github-actions crossbow submit -g r

**Comment by kou** [comment]
@github-actions crossbow submit -g r

**Comment by kou** [comment]
@github-actions crossbow submit -g r

**Comment by kou** [comment]
@github-actions crossbow submit -g r

**Comment by kou** [comment]
@github-actions crossbow submit -g r

**Comment by kou** [comment]
@github-actions crossbow submit -g r

**Comment by kou** [comment]
@github-actions crossbow submit -g r

**Comment by kou** [comment]
@github-actions crossbow submit -g r

**Comment by kou** [comment]
@github-actions crossbow submit -g r

**Comment by kou** [comment]
This is ready to review/merge. Failed jobs are unrelated to this change.

@thisisnic Could you test the latest change to confirm that this approach fixes the original problem too before we merge this?

**Comment by kou** [comment]
I'll merge this in the next week if nobody objects it.

File: r/configure [hunk 2]
```diff
@@ -149,53 +150,42 @@ else
...
       if [ "${ARROW_DEPENDENCY_SOURCE}" = "" ]; then
-        ARROW_DEPENDENCY_SOURCE=AUTO; export ARROW_DEPENDENCY_SOURCE
+        export ARROW_DEPENDENCY_SOURCE=AUTO
       fi
       if [ "${ARROW_DEPENDENCY_SOURCE}" = "AUTO" ]; then
         pkg-config --version >/dev/null 2>&1
         if [ $? -ne 0 ]; then
-          export ARROW_DEPENDENCY_SOURCE=BUNDLED
-          echo "**** pkg-config not installed, setting ARROW_DEPENDENCY_SOURCE=BUNDLED"
+          echo "**** pkg-config is required to use automatically installed Apache Arrow C++"
```
Comment (nealrichardson) [inline r/configure:164]: Why this change?

File: ci/scripts/r_docker_configure.sh [hunk 3]
```diff
@@ -74,20 +74,105 @@ if [[ -n "$DEVTOOLSET_VERSION" ]]; then
   $PACKAGE_MANAGER install -y "devtoolset-$DEVTOOLSET_VERSION"
 fi
 
+case "$PACKAGE_MANAGER" in
```
Comment (nealrichardson) [inline ci/scripts/r_docker_configure.sh:77]: Why does the R setup need Python installed now?

File: dev/tasks/r/github.packages.yml [hunk 4]
```diff
@@ -309,12 +309,22 @@ jobs:
...
+      - name: Install curl and its dependencies for static link
         run: |
-          # zstd is installed by Homebrew on GitHub Actions.
-          echo "PKG_CONFIG_PATH=$(brew --prefix)/lib/pkgconfig${PKG_CONFIG_PATH:+:${PKG_CONFIG_PATH}}" \
-            >> "$GITHUB_ENV"
+          sudo apt update
+          sudo apt install -y \
+            libbrotli-dev \
+            libcurl4-openssl-dev \
+            libidn2-dev \
```
Comment (nealrichardson) [inline dev/tasks/r/github.packages.yml:318]: I don't understand, why do we need all of these extra dependencies now? If they're dependencies of libcurl, why isn't installing libcurl enough?

**Comment by thisisnic** [comment]
> This is ready to review/merge. Failed jobs are unrelated to this change.
> 
> @thisisnic Could you test the latest change to confirm that this approach fixes the original problem too before we merge this?

Sorry, been away for a few days but happy to test this if still needed?  Given a lot of discussions above are unresolved, I'll wait until you've resolved these questions before testing this. Ping me when you want me to look at it again @kou

File: r/configure (hunk 2 above)
Comment (nealrichardson) [reply r/configure:164]: But if dependency source is BUNDLED, we don't need pkg-config because all of the dependencies (aside from curl) are in `-larrow_bundled_dependences`, right?

My question isn't about whether we need pkg-config for source AUTO--that makes sense and was already the case--it is that you changed the default behavior to fail if pkg-config isn't found, rather than just using source BUNDLED.

File: ci/scripts/r_docker_configure.sh (hunk 3 above)
Comment (nealrichardson) [reply ci/scripts/r_docker_configure.sh:77]: Yeah I think I'd like that inside a ARROW_GCS check, just to be clear that that's why we would need Python.

**Comment by assignUser** [comment]
+1

File: ci/scripts/r_docker_configure.sh (hunk 3 above)
Comment (kou) [reply ci/scripts/r_docker_configure.sh:77]: OK. I'll do it.

File: r/configure (hunk 2 above)
Comment (kou) [reply r/configure:164]: Ah, I understand. Sorry. I misunderstood the intent of this code.

I'll revert this change.

**Comment by kou** [comment]
@github-actions crossbow submit -g r

**Comment by amol-** [comment]
@kou is this something still needed? Has been around for a while without any change for more than a month at this point

**Comment by kou** [comment]
Sorry.
I prioritized release related tasks for 10.0.0/10.0.1 and daily maintenance tasks (PR reviews, reply on GitHub/mailing lists and so on) than this.
Most of release related tasks are done now. I'll complete this in a few weeks.

**Comment by kou** [comment]
@github-actions crossbow submit -g r

**Comment by kou** [comment]
@github-actions crossbow submit -g r

**Comment by kou** [comment]
@github-actions crossbow submit -g r

**Comment by kou** [comment]
@github-actions crossbow submit test-r-rstudio-r-base-4.1-opensuse153

**Comment by kou** [comment]
@github-actions crossbow submit test-r-rstudio-r-base-4.1-opensuse153

**Comment by kou** [comment]
@github-actions crossbow submit r-binary-packages

**Comment by kou** [comment]
@github-actions crossbow submit r-binary-packages

**Comment by kou** [comment]
This is ready to merge/review.

@thisisnic Could you test the latest change to confirm that this approach fixes the original problem too before we merge this?

**Review by assignUser — COMMENTED** [review]
some minor comments

**Comment by assignUser** [comment]
I have re-run the failed macos 10.13 jobs. The runner was offline but  is back online again now!

**Comment by kou** [comment]
@github-actions crossbow submit -g r

**Comment by kou** [comment]
@github-actions crossbow submit -g r

File: dev/tasks/r/github.macos-linux.local.yml [hunk 5]
```diff
@@ -45,11 +45,20 @@ jobs:
...
+            libidn2-dev \
+            libkrb5-dev \
+            libldap-dev \
+            libnghttp2-dev \
+            libpsl-dev \
+            librtmp-dev \
+            libssh-dev \
+            libssh2-1-dev \
+            libssl-dev \
+            libzstd-dev
```
[the remark this answers is not in the document]
Comment (kou) [reply dev/tasks/r/github.macos-linux.local.yml]: Good idea. I extracted it as a script in `ci/scripts/` to reduce the duplication in `ci/scripts/r_docker_configure.sh` too.

**Comment by kou** [comment]
@github-actions crossbow submit -g r

**Comment by kou** [comment]
@github-actions crossbow submit -g r

**Comment by kou** [comment]
This is ready again.

**Comment by kou** [comment]
I'll merge this in this week if nobody objects it.

**Comment by thisisnic** [comment]
Thanks @kou and sorry for the delay in responding here - had a complicated week at work and just catching up on my notifications now.

I tried this locally, and looks like it worked fine with my local version of the AWS SDK components, so I'm happy.  I've just pinged Neal to take a final look as he's much more familiar with all of this stuff from a general R package point of view.

**Comment by nealrichardson** [comment]
This feels like a really big change to support something that practically no one has (how many people are building aws-sdk-cpp separately from source--it's not available in package managers so that's what this entails, right?--and then trying to install the R package?). But we don't build with ARROW_S3 on CRAN so I guess the risk there is lower. 

My other concern is that it now looks like we have all of these other system dependencies, and that we're no longer testing more stripped down environments that don't have all of these libraries installed. Do I read that correctly? 

That said, I trust Kou's judgment, so I won't get in the way of merging.

**Comment by kou** [comment]
@github-actions crossbow submit -g r

**Comment by kou** [comment]
@github-actions crossbow submit -g r

**Comment by kou** [comment]
Failures are unrelated because they are happen in nightly CI too.

I merge this.

**Comment by raulcd** [comment]
Are there new dependencies required to build due to this PR? I can see the job for our cookbooks fail if I use the nightlies:
https://github.com/raulcd/arrow-cookbook/actions/runs/3730980375/jobs/6328817362
```
 /usr/bin/ld: cannot find -lnghttp2: No such file or directory
/usr/bin/ld: cannot find -lidn2: No such file or directory
/usr/bin/ld: cannot find -lrtmp: No such file or directory
/usr/bin/ld: cannot find -lssh: No such file or directory
/usr/bin/ld: cannot find -lpsl: No such file or directory
/usr/bin/ld: cannot find -lgssapi_krb5: No such file or directory
/usr/bin/ld: cannot find -llber: No such file or directory
/usr/bin/ld: cannot find -lldap: No such file or directory
/usr/bin/ld: cannot find -llber: No such file or directory
collect2: error: ld returned 1 exit status
```
This PR could be related right? This is the workflow definition: https://github.com/apache/arrow-cookbook/blob/main/.github/workflows/test_r_cookbook.yml#L47

**Comment by raulcd** [comment]
I'll fix it on the cookbooks, I've tested adding the dependencies on a PR I am already working on.

**Comment by assignUser** [comment]
@raulcd  afaik it should also work to add `aws_SOURCE=BUNDLED` (or what ever the var is called) to force building it.

**Comment by kou** [comment]
Oh... I didn't notice that these libraries are required with pre-built Apache Arrow C++ binaries...

It seems that I should implement the following:

https://github.com/apache/arrow/pull/14235#issuecomment-1350558076

> If we use shared libraries, we only need libcurl-dev. We don't need to install additional packages such as libssh-dev because libcurl-dev depends on needed packages for shared linking.

Could you install https://github.com/apache/arrow/pull/14235/files#diff-8a7b064de580a41c3d6212eeeef79930236caf01b58c73407c84fec929b39260R62-R74 for now?
